Add aligned_accessor example

mhoemmen opened the pull request:
1. `aligned_accessor` uses compiler identification macros to pick from various, possibly nonstandard ways to mark a pointer (its value, and/or its type) as having a specific alignment.

2. "Aligned" benchmarks use overaligned allocation and the aforementioned way to mark a pointer as aligned.

3. "Unaligned' benchmarks access memory aligned only to the element type, by adding 1 (`sizeof(ElementType)` bytes) to the array start.

mhoemmen committed:
Add aligned_accessor example

aligned_accessor uses compiler identification macros to pick
from various, possibly nonstandard ways to mark a pointer
(its value, and/or its type) as having a specific alignment.

"Aligned" benchmarks use overaligned allocation
and the aforementioned way to mark a pointer as aligned.

"Unaligned' benchmarks access memory aligned only
to the element type, by adding 1 (sizeof(ElementType) bytes)
to the array start.

If building with OpenMP (i.e., if _OPENMP is defined),
more benchmarks will build and run.  All of them use
"#pragma omp simd", and some add an "aligned" clause.

The example builds with C++14, C++17, and C++20.
I've tested it with the following compilers and options:

* x86-64 GCC 12.1: -std=c++2b -O3 -Wall
* x86-64 clang 14.0.0: -std=c++14 -O3 -Wall
* x86-64 icc 2021.5.0: -std=c++17 -O3 -qopenmp
* x86 msvc v19.32: /std:c++latest
* x86-64 GCC 12.1: -std=c++14 -Wall
* x86-64 GCC 12.1: -std=c++20 -Wall
* x86-64 GCC 9.4.0: -std=c++17

mhoemmen commented:
Fixes for Clang, MSVC, and Intel here: https://godbolt.org/z/vhGz3zcYn

I still need to incorporate these into the PR.

mhoemmen commented:
@crtrott @dalg24 @nliber I updated the PR to fix Clang, MSVC, and Intel builds.  Here is the corresponding Compiler Explorer link: https://godbolt.org/z/47dEf59rs

mhoemmen commented:
Test with C++14, C++17, and C++20 builds: https://godbolt.org/z/raYsGTq3P

mhoemmen commented:
Latest update (just pushed) includes OpenMP fixes: https://godbolt.org/z/PbMWbKMj7

dalg24 commented on examples/aligned_accessor/aligned_accessor.cpp:
```diff
@@ -0,0 +1,714 @@
+valid_byte_alignment(const std::size_t byte_alignment,
+		     const std::size_t size_of_element)
+{
+  return byte_alignment >= alignof(ElementType); // && is a power of two
```
compare to `bit_floor(byte_alignment)` when C++20 is available?

mhoemmen replied:
@dalg24 Excellent suggestion; thank you!  I should be able to use the feature test macro `__cpp_lib_int_pow2` for these C++20 functions.

mhoemmen replied:
Fixed; thank you!

dalg24 commented on examples/aligned_accessor/aligned_accessor.cpp:
```diff
@@ -0,0 +1,714 @@
+  return byte_alignment >= alignof(ElementType); // && is a power of two
+}
+
+template<class ElementType, std::size_t byte_alignment>
+struct aligned_pointer {
+  // x86-64 ICC 2021.5.0 emits warning #3186 ("expected typedef declaration") here.
+  // No other compiler (including Clang, which has a similar type attribute) has this issue.
+  using type = ElementType* _MDSPAN_ALIGN_VALUE_ATTRIBUTE( byte_alignment );
+};
+
+template<class ElementType, std::size_t byte_alignment>
+using aligned_pointer_t = typename aligned_pointer<ElementType, byte_alignment>::type;
```
Why do you bother with the trait and don't directly define the alias?

mhoemmen replied:
Good point -- I must have mixed up in my head some missing C++14 thing from this C++11 feature.  I'll change it.

mhoemmen replied:
Fixed; thanks!

mhoemmen replied:
I ended up bringing back the struct, because this lets me check alignment at compile time for all uses of the alias.  (Just defining the alias directly wouldn't let me impose a `static_assert` on it for all specializations.)

dalg24 commented on examples/aligned_accessor/aligned_accessor.cpp:
```diff
@@ -0,0 +1,714 @@
+
+  constexpr aligned_accessor() noexcept = default;
+
+  MDSPAN_TEMPLATE_REQUIRES(
+    class OtherElementType,
+    std::size_t other_byte_alignment,
+    /* requires */ (std::is_convertible<OtherElementType(*)[], element_type(*)[]>::value && other_byte_alignment == byte_alignment)
+    )
+  constexpr aligned_accessor(aligned_accessor<OtherElementType, other_byte_alignment>) noexcept {}
```
Why a requires clause rather than a static_assert?

mhoemmen replied:
@dalg24 The `element_type` part of this imitates `default_accessor`.  It felt weird to treat byte alignment differently than `element_type`.  I won't insist, though ; - ) .

The byte alignment part could, in theory, permit conversions to lesser alignment.

mhoemmen replied:
@dalg24 Please let me know if you still want this changed; thanks!

dalg24 commented on examples/aligned_accessor/aligned_accessor.cpp:
```diff
@@ -0,0 +1,714 @@
+			aligned_pointer_t<const ElementType, byte_alignment> x,
+			aligned_pointer_t<const ElementType, byte_alignment> y,
+			aligned_pointer_t<ElementType, byte_alignment> z,
+			std::integral_constant<std::size_t, byte_alignment> /* ba */ )
```
Why do you bother with that argument

mhoemmen replied:
1. Consistency with the aligned mdspan interface
2. `aligned_pointer_t<ElementType, byte_alignment` is an alias and might just be `ElementType*`, but it might be a pointer with an attribute.  I wasn't sure whether compilers could deduce `byte_alignment` from an attribute on the pointer.

I'll experiment with removing this parameter.

mhoemmen replied:
I was able to remove this parameter; thanks!

dalg24 reviewed (commented):
A few drive by comments

mhoemmen committed:
aligned_accessor example: improvements

1. Remove unused parameter of valid_byte_alignment

2. Make valid_byte_alignment use C++20 function has_single_bit
   if available

3. Make aligned_pointer_t use valid_byte_alignment to check alignment
   at compile time.  valid_byte_alignment was unused before.

4. Remove unnecessary parameter from add_aligned_raw_1d.

5. Remove unneeded casts and assignments.

dalg24 commented on examples/aligned_accessor/aligned_accessor.cpp:
```diff
@@ -0,0 +1,737 @@
+using test_value_type = float;
+constexpr int min_overalignment_factor = 8;
+constexpr int min_byte_alignment =
+  static_cast<int>(min_overalignment_factor * sizeof(float));
```
cast is pointless

mhoemmen replied:
@dalg24 Fixed; thanks!

dalg24 commented on examples/aligned_accessor/aligned_accessor.cpp:
```diff
@@ -0,0 +1,737 @@
+
+#define TICK() const auto tick = std::chrono::steady_clock::now()
+
+using double_seconds = std::chrono::duration<double, std::ratio<1>>;
```
Why would you define an alias if you ever only use it in the TOCK macro definition?

mhoemmen replied:
It makes the definition of `TOCK` shorter and better expresses the code's intent.

dalg24 commented on examples/aligned_accessor/aligned_accessor.cpp:
```diff
@@ -0,0 +1,737 @@
+
+#endif // _OPENMP
+
+  cout << "Number of trials: " << num_trials << endl
```
Flushing the stream excessively

mhoemmen replied:
I'm not too worried about this, since this is the end of the benchmark and not performance-critical code.

dalg24 commented on examples/aligned_accessor/aligned_accessor.cpp:
```diff
@@ -0,0 +1,737 @@
+    return -1;
+  }
+
+  int n = 10000;
+  int num_trials = 100;
```
Why do you bother with initializing to overwrite the value below

mhoemmen replied:
@dalg24 Fixed; thanks!

dalg24 reviewed (commented):
Looks mostly good
Only question I have is do you care to not cast within the outer repetition loop?
Because if that is acceptable you would be able to get rid of a lot of boiler plate code

mhoemmen commented:
@dalg24 wrote:
> Only question I have is do you care to not cast within the outer repetition loop?

The "casts" in the various `benchmark_` functions don't all have the same form.  Some of them actually create an mdspan from a pointer.  Others call `bless` on the pointer, which both casts and calls `assume_aligned`.  Performing the casts or blessings explicitly, outside the outer loop, lets each `add_*` function be as simple as possible.  This hopefully removes any unnecessary barriers to the compiler optimizing it.

This is an example, so the small amount of repetition has educational value.  Users can copy and paste as a start to adding their own benchmarks.

crtrott commented on examples/aligned_accessor/aligned_accessor.cpp:
```diff
@@ -0,0 +1,734 @@
+  //
+  // This exists so that we can compare performance
+  // of aligned and unaligned access.
+  ElementType* data_unaligned() const noexcept
```
Not sure if this is what we should do here instead of returning also address to element 0. After all you are now measuring something different, like the number of cachelines/pages touched by each thread might be different. I'd rather like to see that the attribute in and of itself does something.

mhoemmen replied:
@crtrott My concern there was that the compiler has complete code visibility, so it could deduce that the allocation is aligned.  For the benchmark, I wanted to be able to give it an allocation that I _know_ is aligned to no more than `sizeof(ElementType)`.

mhoemmen replied:
@crtrott I like the tag idea for the benchmarks, though!  I was a bit worried about possibly needing to do type conversion of x, y, and z on every invocation of the inner loop, but I think I could restructure the code to avoid that.

mhoemmen replied:
@crtrott I tried out the tag idea, but I don't think it saves much code.  Each benchmark invocation still needs to name the const and nonconst array types and the tag types explicitly.

Perhaps y'all are missing the point a bit -- this is an _example_.  It's code for users to read, that teaches them how to write a custom accessor.  It's something more like a tutorial.  Reducing code duplication isn't really the point here; the point is simple, readable code that users could easily copy and paste.

crtrott reviewed (changes requested):
I think we need to change the non-aligned element returning & to element 0 too, I am also wondering if we can reduce code duplication here. For example have the same benchmark function templated on a tag and the casted to type:

```c++
template<class ArrayType, class constArrayType, class ElementType,  class BenchmarkTag, class ... ConstructorArgs>
auto benchmark_add_1d(const BenchMarkTag& tag, const std::size_t num_trials,
				      const index_type n,
				      const ElementType x[],
				      const ElementType y[],
				      ElementType z[],
                                      ConstructorArgs ... cargs)
{
  TICK();
  constArrayType x2{x, cargs};
  constArrayType y2{x, cargs};
  ArrayType z2{x, cargs};
  for (std::size_t trial = 0; trial < num_trials; ++trial) {
    add(tag, x2, y2, z2);
  }
  return TOCK();
}
```

mhoemmen committed:
Suppress Intel compiler's spurious warning

mhoemmen committed:
aligned_accessor: Address review feedback

Don't use a single allocation for aligned and unaligned access
any more.  Instead, separate those allocations, and encapsulate
them more.  Add comments to make it clear why we want a
"deliberately unaligned array" for benchmarks and for studying
code generation.  Also, make sure that the elements of the arrays
are filled before we start computing with them.

mhoemmen commented:
@crtrott wrote:
> I think we need to change the non-aligned element returning & to element 0 too....

I have addressed this in my latest revision.
